exercises/binary-search/test/test_binary_search.c

```diff
@@ -0,0 +1,91 @@
+#include "vendor/unity.h"
+#include "../src/binary_search.h"
+
+void test_single_element(void)
+{
+   int arr[] = { 6 };
+   size_t length = sizeof(arr) / sizeof(arr[0]);
+   TEST_ASSERT_EQUAL_UINT64(0, binary_search(6, arr, length));
+}
+
+void test_value_in_middle(void)
+{
+   int arr[] = { 1, 3, 4, 6, 8, 9, 11 };;
+   size_t length = sizeof(arr) / sizeof(arr[0]);
+   TEST_ASSERT_EQUAL_UINT64(3, binary_search(6, arr, length));
+}
+
+void test_value_at_beginning(void)
+{
+   int arr[] = { 1, 3, 4, 6, 8, 9, 11 };;
+   size_t length = sizeof(arr) / sizeof(arr[0]);
+   TEST_ASSERT_EQUAL_UINT64(0, binary_search(1, arr, length));
+}
+
+void test_value_at_end(void)
+{
+   int arr[] = { 1, 3, 4, 6, 8, 9, 11 };;
+   size_t length = sizeof(arr) / sizeof(arr[0]);
+   TEST_ASSERT_EQUAL_UINT64(6, binary_search(11, arr, length));
+}
+
+void test_find_value_with_odd_length(void)
+{
+   int arr[] = { 1, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 634 };
+   size_t length = sizeof(arr) / sizeof(arr[0]);
+   TEST_ASSERT_EQUAL_UINT64(9, binary_search(144, arr, length));
+}
+
+void test_find_value_with_even_length(void)
+{
+   int arr[] = { 1, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377 };
+   size_t length = sizeof(arr) / sizeof(arr[0]);
+   TEST_ASSERT_EQUAL_UINT64(5, binary_search(21, arr, length));
+}
+
+void test_identify_missing_value(void)
+{
+   int arr[] = { 1, 3, 4, 6, 8, 9, 11 };;
+   size_t length = sizeof(arr) / sizeof(arr[0]);
+   TEST_ASSERT_EQUAL_UINT64(length, binary_search(7, arr, length));
+}
+
+void test_value_smaller_than_everything(void)
+{
+   int arr[] = { 1, 3, 4, 6, 8, 9, 11 };;
+   size_t length = sizeof(arr) / sizeof(arr[0]);
+   TEST_ASSERT_EQUAL_UINT64(length, binary_search(0, arr, length));
+}
+
+void test_value_larger_than_everything(void)
+{
+   int arr[] = { 1, 3, 4, 6, 8, 9, 11 };;
+   size_t length = sizeof(arr) / sizeof(arr[0]);
+   TEST_ASSERT_EQUAL_UINT64(length, binary_search(13, arr, length));
+}
+
+void test_empty_array(void)
+{
+   int *arr = NULL;
```

Comment:
The way I see it, `bsearch` has a few different scenarios regarding `NULL` and whether or not it's actually valid input.

1. `ptr = NULL, length = 0` - **Valid** - Returns not found.
2. `ptr =~NULL, length = 0` - **Valid** - Returns not found.
3. `ptr = NULL, length = ~0` - **Invalid** - Undefined behaviour.

The second scenario I should probably add a test for, but I think we can call the 3rd undefined behaviour and not handle it.